### PR TITLE
Map Template Load Verb + Maploader Improvements

### PR DESCRIPTION
**Pull request by RemieRichards**
## Map Templates:
- Admins with +DEBUG can now Load map files in the <B>_map/templates</B> folder where they (or their ghost) is standing.
- Admins with +DEBUG can now Upload a map file (smaller than 1MB) to the server for use within that round, it does not persist between rounds.

Includes two examples:
- <B>small_asteroid_1:</B> an asteroid with low chance for minerals
- <B>small_shuttle_1:</B> a small shuttle with some chairs

Example naming scheme: <B>"size_quick-description-or-title_instance-number"</B>
I'd like this naming scheme to be followed, but it's not necessary
(I might swap size and title around, so "asteroid_small" appears at the top)

example
## Maploader Improvements:
- Maploader `for(X = 1, X<=some_max, X++)` loops replaced with the much faster `for(X in 1 to some_max)` format, removes the 1-2ms delay after loading a maptemplate, woo!
- Maploader no longer assumes that the coordinates it is given exist, this has NOT slowed down the maploader, _AND_ prevents runtimes from loading templates at map edge, go ahead, load half a shuttle off the z level, the game doesn't give a -shit- anymore!

### CODE REVIEW

**Comment by AnturK** [comment]
What with lava ruins and such i think all these should get unified datums that preload their properties. But i guess that can wait.

**Comment by octareenroon91** [comment]
What would you want datumized?  The main drawback I see is that your areas will be the same in each 'paste' of the map.

**Comment by RemieRichards** [comment]
@octareenroon91 with a datum for each "paste" you could "un-load" them, but I dont think that's that useful right now.

**Comment by octareenroon91** [comment]
Ah, very true.

**Comment by Niknakflak** [comment]
oooohhhhh shiny

**Comment by MMMiracles** [comment]
Is there any limitation reasoning why a system similar to midi uploading be used to upload your own custom .dmm past invalid paths? Most station maps average around 1.5MB, with meta being the biggest at around 2MB so I'd assume uploading little segments like this wouldn't be that intensive.

**Comment by RemieRichards** [comment]
@MMMiracles literally just because I don't know the exact procs for sending something to the server (things like VVing icon just use "as icon" which apparently defaults to the user's PC first, which I guess automatically does the transfer, I *THINK* there's an "as file" in which case I'd just ban non .dmm files and that'd be that)

tl;dr I'll do it if I can figure it out.

**Comment by Jordie0608** [comment]
@RemieRichards 
https://github.com/tgstation/-tg-station/blob/master/code/modules/client/client%20procs.dm#L80
AllowUpload() gets called everytime a client tries to send a file via input()

**Comment by RemieRichards** [comment]
@Jordie0608 ah excellent! 
I don't suppose there's a way to determine where it ends up? although I suppose we could leave _maps/templates as a "curated" folder, and uploads as just w/e the admin feels at the time.

the 1MB limit should be fine too, as previously mentioned, the "small shuttle" here is only 1.9 kilobytes.
(the asteroid isn't even half a kilobyte (219 Bytes)

**Comment by Jordie0608** [comment]
<Jordie> mso files uploaded to the server go to byond\cache folder yes?
<MrStonedOne> yes
<MrStonedOne> then they get merged into the .dyn.rsc

**Comment by RemieRichards** [comment]
@Jordie0608 I'm going to assume that's and IRC log dump, and not you going insane :P

**Comment by Jordie0608** [comment]
Apparently < and > are markdown tags

**Comment by Bawhoppen** [comment]
hell yes.

File: code/modules/admin/verbs/map_template_loadverb.dm [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+
+/client/proc/map_template_upload()
+	set category = "Debug"
+	set name = "Map Template - Upload"
+
+	var/map = input(usr, "Choose a Map Template to upload to template storage","Upload Map Template") as null|file
+	if(!map)
+		return
+	if(!findtext("[map]",".dmm"))
```
Comment (tkdrg) [inline code/modules/admin/verbs/map_template_loadverb.dm]: Can't you use `copytext` to make sure it isn't something like `innocent_map.dmm.mp4.swf.exe`

Comment (RemieRichards) [inline code/modules/admin/verbs/map_template_loadverb.dm]: I suppose so.

**Comment by Bawhoppen** [comment]
also I just thought, could this work for holodeck/thunderdome templates?
and if so, should it?

**Comment by octareenroon91** [comment]
It would make the holodeck system easier to maintain since you don't have to map onto z2.dmm .
I don't foresee any major difference in resource usage though.
